File: Python/v38/library.py
class Item:    

    def __init__(self, title, author, year):
        self._title = title
        self._author = author
        self._year = year
        self._status = ""

    def display_info(self):
        pass

    def set_status(self, status):
        self._status = status

class Book(Item):

    def __init__(self, title, author, year, pages):
        self._pages = pages
        super().__init__(title, author, year)

    def display_info(self):
        print(f"Title: {self._title}\nAuthor: {self._author}\nYear: {self._year}\nPages: {self._pages}")

class Magazine(Item):

    def __init__(self, title, author, year, publication):
        self._publication = publication
        super().__init__(title, author, year)

    def display_info(self):
        print(f"Title: {self._title}\nAuthor: {self._author}\nYear: {self._year}\nPublication: {self._publication}")


class LibraryUser:
    def __init__(self, name):
        self._name = name
        self._borrowed_items = []
    
    def borrow(self, Item):
        self._borrowed_items.append(Item)
    
    def return_item(self, Item):
        self._borrowed_items.remove(Item)

class Library:
    def __init__(self):
        self._items = []
        self._users = []

    def add_item(self, Item):
        self._items.append(Item)

    def remove_item(self, Item):
        self._items.remove(Item)

    def register_user(self, LibraryUser):
        self._users.append(LibraryUser)

    def borrow_item(self, LibraryUser, Item):
        if Item in self._items:
            if Item._status == "":
                Item.set_status(LibraryUser._name)
                LibraryUser.borrow(Item)
            else:
                print(f"Not possible, {Item._title} is borrowed by {Item._status}") 
        else:
            print(f"Not possible, {Item._title} does not exist.")

    def return_item(self, LibraryUser, Item):
        Item.set_status("")
        LibraryUser.return_item(Item)
    
    def list_items(self):        
        for item in self._items:
            item.display_info()            
            print("---------")

    def list_available_items(self):
        for item in self._items:
            if item._status == "":
                item.display_info()            
                print("---------")

    def summary(self):
        for item in self._items:
            if item._status != "":
                print(f"{item._title} is borrowed by {item._status}")
            else:
                print(f"{item._title} is available")


""" library = Library()
user1 = LibraryUser("Pelle")
user2 = LibraryUser("Emma")

library.register_user(user1)
library.register_user(user2)

book1 = Magazine("News", "Pelle", 2005, "Street")
book2 = Book("Swan Lake", "Emma", 1999, 500)
book3 = Book("404", "null", 000, 0)

library.add_item(book1)
library.add_item(book2)

print("---------")
library.list_available_items()

library.remove_item(book1)
print("---------")
library.list_available_items()

library.add_item(book1)

library.borrow_item(user1, book2)
print("---------")
print("Available items")
library.list_available_items()
print("All items")
library.list_items()

library.borrow_item(user2, book2)
library.borrow_item(user2, book3)
library.summary() """

File: C/v45/library.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <memory>

class Item {
    protected:
        std::string title;
        std::string author;
        int year;

    public:
        Item(std::string title, std::string author, int year) : title(title), author(author), year(year) {}
        virtual void display_info() const {};
        virtual ~Item() {};
        bool operator==(const Item& other) const {
            return title == other.title && author == other.author && year == other.year;        
        }
        std::string get_title() const {
            return title;
        }
};

class Book : public Item {
    protected:
        int pages;
    
    public:
        Book(std::string title, std::string author, int year, int pages)
             : Item(title, author, year), pages(pages) {}
        void display_info() const override {
            std::cout << title << ", " << author << ", " << year << ", " << pages << std::endl; 
        }
};

class Magazine : public Item {
    protected:
        int issue;

    public:
        Magazine(std::string title, std::string author, int year, int issue)
             : Item(title, author, year), issue(issue) {}
        void display_info() const override {
                std::cout << title << ", " << author << ", " << year << ", " << issue << std::endl; 
        }
};

class LibraryUser {
    protected:
        std::string name;
        std::vector<Item*> borrowed_items;

    public:
        LibraryUser(std::string name) : name(name) {}
        void borrow(Item& item) {
            borrowed_items.push_back(&item);            
        }
        void return_item(const Item& item) {
            auto it = std::remove_if(borrowed_items.begin(), borrowed_items.end(),
                [&item](Item* i) {
                    return i->get_title() == item.get_title();
                });
            if (it != borrowed_items.end()) {
                borrowed_items.erase(it, borrowed_items.end());
            }
        }
        void list_borrowed_items() const {
            for (const Item* item : borrowed_items) {
                item->display_info();
            }
        }
        std::string get_name() const {
            return name;
        }
};

class Library {
    protected:
        std::vector<Item*> items;
        std::vector<LibraryUser*> users;

    public:
        void add_item(Item& item) {
            items.push_back(&item);
        }
        void remove_item(const Item& item) {
            auto it = std::remove_if(items.begin(), items.end(),
                [&item](Item* i) {
                    return i->get_title() == item.get_title();
                });
            if (it != items.end()) {
                items.erase(it, items.end());
            }
        }
        void register_user(LibraryUser& user) {
            users.push_back(&user);
        }
        void borrow_item(LibraryUser& user, Item& item) {
            user.borrow(item);
            remove_item(item);
        }
        void return_item(LibraryUser& user, Item& item) {
            user.return_item(item);
            add_item(item);
        }
        void list_items() const {            
            std::cout << "In library: " << std::endl;
            list_available_items();
            std::cout << "Borrowed by users: " << std::endl;
            for (const LibraryUser* user : users) {
                user->list_borrowed_items();
            }
        }
        void list_available_items() const {
            for(const Item* item : items) {
                item->display_info();
            }
        }
        void list_borrowed_items() const {
            for(const LibraryUser* user: users) {
                std::cout << "------------------" << std::endl;
                std::cout << "Books borrowed by " << user->get_name() << std::endl;
                user->list_borrowed_items();
            }
        }
};


int main() {
    Library library;
    
    LibraryUser pelle = LibraryUser("Pelle");
    LibraryUser lisa = LibraryUser("Lisa");
    library.register_user(pelle);
    library.register_user(lisa);
    

    Magazine mag = Magazine("Magazine", "Anna", 2020, 5);
    Book book = Book("Book", "Carl", 1990, 500);

    library.add_item(mag);
    library.add_item(book);

    std::cout << "Test list_available_items()\n";
    library.list_available_items();

    std::cout << "Test all items\n";
    library.borrow_item(pelle, mag);
    library.list_items();

    std::cout << "Test give back item\n";
    library.return_item(pelle, mag);
    library.list_items();

    std::cout << "Test list of borrowed books\n";
    library.borrow_item(pelle, book);
    library.borrow_item(lisa, mag);
    library.list_borrowed_items();

    return 0;
}